abi/src/specs.rs

```diff
@@ -190,6 +190,7 @@ pub struct ConstructorSpec<F: Form = MetaForm> {
     /// The name of the message.
     name: F::String,
     /// The selector hash of the message.
+    #[serde(serialize_with = "serialize_selector")]
     selector: u32,
```

Comment:
This solution is not bad since it is kept pretty local. However, what we really want is to have `[u8; 4]` at this position fed by `ink_lang` correctly. With this PR we would have the situation of `ink_lang` converting its `[u8; 4]` magically into a `u32` just so that we can finally convert it back into `[u8; 4]` before serializing. This is also dangerous since nothing guarantees us from `ink_lang` how the conversion from `[u8; 4]` actually works.

Reply:
Ok, so to make sure I understand you correctly: What you suggest is to instead replace the `ink_lang` internal `selector` type `u32` with `[u8; 4]` everywhere, correct?

Reply:
Yes but this is not a local change as your current PR is. So I can totally understand why you went this way. However, this double transformation that we have with this PR is definitely not something I want to elaborate further. So let us put in a bit more work and fix it properly. :)